bin/propolis-server/src/lib/vm/state_driver.rs

```diff
@@ -524,32 +535,208 @@ impl StateDriver {
+        // e.g. due to sled failure, so these configurations aren't necessarily
+        // client errors.)
+        //
+        // Before getting into any of that, handle the synchronous portions of
+        // VM startup. First, ensure that the kernel VM and all its associated
+        // devices are in the correct initial states.
+        let objects = self.objects.lock_shared().await;
+        match start_reason {
+            // If this VM is a migration target, migration will have properly
+            // initialized the vCPUs, but will have left the kernel VM paused.
+            // Resume it here before asking any in-kernel components to start.
+            VmStartReason::MigratedIn => objects.resume_kernel_vm(),
+
+            // If this VM is starting from scratch, its kernel VM is active, but
+            // its vCPUs have not been initialized yet.
+            VmStartReason::ExplicitRequest => objects.reset_vcpus(),
+        }
 
-        match &start_result {
-            Ok(()) => {
-                self.external_state.update(ExternalStateUpdate::Instance(
-                    InstanceState::Running,
-                ));
+        // Send synchronous start commands to all devices.
+        for (name, dev) in objects.device_map() {
+            info!(self.log, "sending start request to {}", name);
+            let res = dev.start();
+            if let Err(e) = res {
+                error!(self.log, "startup failed for {}: {:?}", name, e);
+                return VmStartOutcome::Failed;
             }
-            Err(e) => {
-                error!(&self.log, "failed to start devices";
-                                 "error" => ?e);
-                self.external_state.update(ExternalStateUpdate::Instance(
-                    InstanceState::Failed,
-                ));
+        }
+
+        // Next, prepare to start block backends. This is done by capturing the
+        // current block backend set and creating a future that issues all the
+        // start requests.
+        //
+        // For this to work, the set of block backends to be started must not
+        // change while the VM is starting. This is guaranteed because all such
+        // requests to hotplug a block backend will be dispatched to the VM's
+        // request queue; if any such requests are seen below, they can simply
+        // be buffered and handled after the rest of the VM has started.
+        async fn start_block_backends(
+            log: slog::Logger,
+            backends: BlockBackendMap,
+        ) -> anyhow::Result<()> {
+            for (name, backend) in backends {
+                info!(log, "starting block backend {}", name);
+                let res = backend.start().await;
+                if let Err(e) = &res {
+                    error!(
+                        log,
+                        "startup failed for block backend {}: {:?}", name, e
+                    );
```

Comment:
similarly, maybe we should use a structured slog field for the block backend name on these logs?